### PR TITLE
feat: split occ endpoints configuration to multiple libraries

### LINKED ISSUE #9112: Split the Occ endpoints config interface in between feature libs
The core config interface `OccEndpoints` (`projects\core\src\occ\occ-models\occ-endpoints.model.ts`) contains a complete list of endpoints, including endpoints needed only by the feature libs. The file is already very busy. For readability and scalability it would be good to split this interface in between the feature libs.
Currently it contains list of endpoints used in the `@spartacus/my-company/organization` lib. See the part of the file:
https://github.com/SAP/spartacus/commit/bc69ae2e86bbe45e17df67f8081a92e60f0923ea#diff-5798d8e2470ec95ac89c9df1309b3b6dR375

It's going to be similar case for the upcoming configurable product feature.

In the end of the day the main purpose of this interface is to provide intellisense code suggestions for customers when configuring Spartacus app.

We need to investigate how to achieve splitting this interface. Maybe _type augmentation_ can help.

### PR DESCRIPTION
**Pull request by LTiger14**
Closes #9112

### CODE REVIEW

File: projects/core/src/occ/config/occ-config-validator.ts [hunk 1]
```diff
@@ -8,4 +8,7 @@ export function occConfigValidator(config: OccConfig) {
   ) {
     return 'Please configure backend.occ.baseUrl before using storefront library!';
   }
+  if (!config.backend?.occ?.prefix?.startsWith('/')) {
```
Comment (marlass) [inline projects/core/src/occ/config/occ-config-validator.ts]: Validator should never change the configuration. I would move it to private methods in `OccSitesConfigLoader` and `OccEndpointsService`. Maybe later `OccEndpointsService` should use `OccEndpointsService` for utilities? Just a thought

File: projects/core/src/occ/adapters/asm/occ-asm.adapter.spec.ts [hunk 2]
```diff
@@ -47,7 +51,11 @@ class MockBaseSiteService {
 }
 
 class MockOccEndpointsService {
```
Comment (marlass) [inline projects/core/src/occ/adapters/asm/occ-asm.adapter.spec.ts]: Would be good to add `implements Partial<OccEndpointsService>` to make future refactoring easier.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 3]
```diff
@@ -50,6 +64,22 @@ export class OccEndpointsService {
   }
 
   /**
+   * Returns the value configured for a specific endpoint
+   *
+   * @param endpointKey the configuration key for the endpoint to return
+   */
+  getRawEndpointValue(endpoint: string, scope?: string): string {
+    if (!this.config?.backend?.occ) {
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: Why we need to check this property here? It doesn't seem related to raw endpoint value.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 4]
```diff
@@ -58,52 +88,116 @@ export class OccEndpointsService {
...
-      this.activeBaseSite
-    );
+    if (!endpointComponents) {
+      return (
+        (this.config.backend.occ.baseUrl || '') +
+        this.config.backend.occ.prefix +
+        this.activeBaseSite
+      );
+    } else {
+      return (
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: Can we have an utility that takes care about all the `/` between segments?
Something similar to `path.join`?

**Review by marlass — CHANGES_REQUESTED** [review]
I still find name of methods in `OccEndpointsService` not descriptive. Without looking at implementation it's hard to guess which one you should use.

I'm wondering if we should not set `prefix: true` and `baseUrl: true` and `baseSite: true` as default. Isn't it the most common case?

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 3 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: You are right, removed

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 5]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
   }
 
   /**
-   * Returns base OCC endpoint (baseUrl + prefix + baseSite)
+   * Returns base OCC endpoint (baseUrl + prefix + baseSite) base on provided values
+   *
+   * @param propertiesToOmit Specify properties to not add to the url (baseUrl, prefix, baseSite)
    */
-  getBaseEndpoint(): string {
+  getOccBaseUrl(propertiesToOmit?: BaseOccUrlProperties): string {
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: I would drop the `Occ` from name.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 6]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
     if (!this.config?.backend?.occ) {
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: I think there is still valid case even if you don't provide it. We can still return url eg. `/electronics-spa` which still can be used.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 7]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
     }
 
-    return (
-      (this.config.backend.occ.baseUrl || '') +
-      this.config.backend.occ.prefix +
-      this.activeBaseSite
-    );
+    if (!propertiesToOmit) {
+      return urlPathJoin(
+        this.config.backend.occ.baseUrl || '',
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: The config handles undefined values right?

```suggestion
        this.config.backend.occ.baseUrl,
```

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 8]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
+        this.config.backend.occ.prefix,
+        this.activeBaseSite
+      );
+    } else {
+      const baseUrl =
+        propertiesToOmit.baseUrl === false
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: I think it would be cleaner if we would add default value in function params and handle here `undefined` the same way we handle `false` value.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 9]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
-   * @param endpoint Endpoint suffix
+   * Add the base OCC url properties to the specified endpoint string
+   *
+   * @param endpointString String value for the url endpoint
+   * @param propertiesToOmit Specify properties to not add to the url (baseUrl, prefix, baseSite)
    */
-  getEndpoint(endpoint: string): string {
-    if (!endpoint.startsWith('/')) {
-      endpoint = '/' + endpoint;
+  buildUrlFromEndpointString(
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: Do we need this function in public API?

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 10]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
+  }
+
+  /**
+   * Returns a fully qualified OCC Url
+   *
+   * @param endpoint Name of the OCC endpoint key
+   * @param attributes Dynamic attributes used to build the url
+   * @param propertiesToOmit Specify properties to not add to the url (baseUrl, prefix, baseSite)
+   */
+  getOccUrlFromConfiguration(
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: What do you think about `buildUrl` name? It is used very often and short name would definitely help and the role is the same as for previous `getUrl`.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 11]
```diff
@@ -58,52 +87,148 @@ export class OccEndpointsService {
...
+    endpoint: string,
+    attributes?: DynamicAttributes,
+    propertiesToOmit?: BaseOccUrlProperties
+  ): string {
+    endpoint = this.getEndpointForScope(endpoint, attributes?.scope);
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: Could we avoid in new methods mutating passed parameters?

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 12]
```diff
@@ -183,4 +313,13 @@ export class OccEndpointsService {
         : endpointConfig?.[DEFAULT_SCOPE]) || endpoint
     );
   }
+
+  private validatePrefixConfig() {
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]: I would avoid mutations on configuration. Let's create method that returns perfix with `/` and replace every usage of `this.config.backend.occ.prefix` with it. Similarly in the other class.

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 7 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: Yes it does

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 12 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: I do agree on this, mutating it feels dirty.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 10 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: Yes that's good. I see the value in the simpler name especially as it's the method that will be the most used of the service.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 9 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: It's used instead of getEndpoint for cases where the configurable endpoints are not used.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 6 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: I needed to keep this for backwards compatibility. I removed it from `getBaseUrl` and put it back in `getBaseEndpoint`.

File: projects/core/src/config/utils/dynamic-template.ts [hunk 13]
```diff
@@ -1,10 +1,24 @@
...
+    encodeVariable?: boolean
+  ): string {
     for (const variableLabel of Object.keys(templateVariables)) {
       const placeholder = new RegExp('\\${' + variableLabel + '}', 'g');
       templateString = templateString.replace(
         placeholder,
-        templateVariables[variableLabel]
+        // TODO 4.0: default to encodeVariable = true
+        encodeVariable
+          ? encodeURIComponent(templateVariables[variableLabel])
```
Comment (marlass) [inline projects/core/src/config/utils/dynamic-template.ts:20]: Should we also apply this change in CDS `DynamicTemplate`?

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 14]
```diff
@@ -50,6 +65,19 @@ export class OccEndpointsService {
   }
 
   /**
+   * Returns the value configured for a specific endpoint
+   *
+   * @param endpointKey the configuration key for the endpoint to return
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts:70]: Add second param in tsdoc

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 15]
```diff
@@ -58,52 +86,138 @@ export class OccEndpointsService {
...
-      this.config.backend.occ.baseUrl +
-      this.config.backend.occ.prefix +
-      endpoint
-    );
+    return this.getEndpoint(endpoint, { baseSite: false });
   }
 
   /**
-   * Returns base OCC endpoint (baseUrl + prefix + baseSite)
+   * @Deprecated since 3.2 - use "buildOccBaseUrl" with the same parameters
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]:
```suggestion
   * @Deprecated since 3.2 - use "getBaseUrl" with the same parameters
```

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 16]
```diff
@@ -58,52 +86,138 @@ export class OccEndpointsService {
...
+  ): string {
+    const baseUrl =
+      baseUrlProperties.baseUrl === false
+        ? ''
+        : this.config.backend.occ.baseUrl;
+    const prefix = baseUrlProperties.prefix === false ? '' : this.getPrefix();
+    const baseSite =
+      baseUrlProperties.baseSite === false ? '' : this.activeBaseSite;
+
+    return urlPathJoin(baseUrl, prefix, baseSite);
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts:148]: I think it will fail in one case. When we want only baseSite url it won't start with `/` and I would assume that this method either returns url starting with `/` or starting with `http`.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 9 above)
Comment (marlass) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: I think we should discourage usage of endpoints that are not configurable and in 4.0 also switch usage of those endpoints to configurable ones.

And with that I think we should make this method private (to simplify transition) and then even remove it.

File: projects/core/src/occ/services/occ-endpoints.service.spec.ts [hunk 17]
```diff
@@ -202,4 +209,153 @@ describe('OccEndpointsService', () => {
       );
     });
   });
+
+  describe('getOccBaseUrl', () => {
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.spec.ts]:
```suggestion
  describe('getBaseUrl', () => {
```

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 9 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts]: Agreed, let's force people to move to configurable endpoints. I'll open a ticket to not forget.

File: projects/core/src/config/utils/dynamic-template.ts (hunk 13 above)
Comment (LTiger14) [reply projects/core/src/config/utils/dynamic-template.ts:20]: I noticed CDS has their own emplementation of the `occ-endpoints.service`. I'd want to clarify why before changing anything in CDS. I would like to move this to a separate issue.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 16 above)
Comment (LTiger14) [reply projects/core/src/occ/services/occ-endpoints.service.ts:148]: With the current implementation it won't add a `/` to the beginning of your url, it will simply join the parts. That said if you give it only a baseSite with this value `/base-site` it will return it as is but if you don't pass a value with a `/`, it won't add it.

To be honest I don't  really see the value in using this method if you only want to get the base site. If we want to start adding `/` at the start for certain conditions it will add a bit of complexity to the method.

File: projects/core/src/config/utils/dynamic-template.ts (hunk 13 above)
Comment (marlass) [reply projects/core/src/config/utils/dynamic-template.ts:20]: Ok. Let's create a ticket for that.

File: projects/core/src/occ/services/occ-endpoints.service.ts (hunk 16 above)
Comment (marlass) [reply projects/core/src/occ/services/occ-endpoints.service.ts:148]: Ok. Let's leave it as it is. If someone reports this use case in the future we will fix it.

File: projects/core/src/occ/services/occ-endpoints.service.ts [hunk 18]
```diff
@@ -58,52 +87,125 @@ export class OccEndpointsService {
...
-    return (
-      (this.config.backend.occ.baseUrl || '') +
-      this.config.backend.occ.prefix +
-      this.activeBaseSite
-    );
+    return this.getBaseUrl(propertiesToOmit);
   }
 
   /**
+   * @Deprecated since 3.2 - use "buildUrlFromEndpointString" with the same parameters
```
Comment (marlass) [inline projects/core/src/occ/services/occ-endpoints.service.ts]:
```suggestion
   * @Deprecated since 3.2 - use "buildUrl" with configurable endpoints instead
```

**Review by marlass — CHANGES_REQUESTED** [review]

**Review by marlass — APPROVED** [review]
